### PR TITLE
cmake: Generate cvc4_export.h and set visibility to hidden.

### PR DESCRIPTION
**Pull request by mpreiner**
cmake will automatically generate an export header cvc4_export.h, which makes sure that the correct export features are
defined depending on the compiler and target platform. The macro CVC4_EXPORT replaces CVC4_PUBLIC and its usage is reduced by 2/3. Visibility is set to `default` for debug builds.

Fixes https://github.com/CVC4/cvc4-projects/issues/96

### CODE REVIEW

File: src/options/mkoptions.py [hunk 1]
```diff
@@ -168,23 +168,23 @@
 
 
 TPL_OPTION_STRUCT_RW = \
-"""extern struct CVC4_PUBLIC {name}__option_t
+"""extern struct  {name}__option_t
```
Comment (nafur) [inline src/options/mkoptions.py:171]:
```suggestion
"""extern struct {name}__option_t
```

File: src/options/mkoptions.py [hunk 2]
```diff
@@ -168,23 +168,23 @@
...
 {{
   typedef {type} type;
   type operator()() const;
   bool wasSetByUser() const;
   void set(const type& v);
   const char* getName() const;
-}} thread_local {name} CVC4_PUBLIC;"""
+}} thread_local {name} ;"""
```
Comment (nafur) [inline src/options/mkoptions.py:178]:
```suggestion
}} thread_local {name};"""
```

File: src/options/mkoptions.py [hunk 3]
```diff
@@ -168,23 +168,23 @@
...
 
 TPL_OPTION_STRUCT_RO = \
-"""extern struct CVC4_PUBLIC {name}__option_t
+"""extern struct  {name}__option_t
```
Comment (nafur) [inline src/options/mkoptions.py:181]:
```suggestion
"""extern struct {name}__option_t
```

File: src/options/mkoptions.py [hunk 4]
```diff
@@ -168,23 +168,23 @@
...
 {{
   typedef {type} type;
   type operator()() const;
   bool wasSetByUser() const;
   const char* getName() const;
-}} thread_local {name} CVC4_PUBLIC;"""
+}} thread_local {name} ;"""
```
Comment (nafur) [inline src/options/mkoptions.py:187]:
```suggestion
}} thread_local {name};"""
```

File: src/options/mkoptions.py [hunk 5]
```diff
@@ -258,9 +258,9 @@
 TPL_DECL_MODE_FUNC = \
 """
 std::ostream&
-operator<<(std::ostream& os, {type} mode) CVC4_PUBLIC;"""
+operator<<(std::ostream& os, {type} mode) ;"""
```
Comment (nafur) [inline src/options/mkoptions.py:261]:
```suggestion
operator<<(std::ostream& os, {type} mode);"""
```

File: src/options/mkoptions.py [hunk 6]
```diff
@@ -258,9 +258,9 @@
...
 
-TPL_IMPL_MODE_FUNC = TPL_DECL_MODE_FUNC[:-len(" CVC4_PUBLIC;")] + \
+TPL_IMPL_MODE_FUNC = TPL_DECL_MODE_FUNC[:-len(" ;")] + \
```
Comment (nafur) [inline src/options/mkoptions.py:263]: Not sure what this is doing, but this seems wrong now.

**Review by nafur — APPROVED** [review]
LGTM, just a few minor formatting issues.
Do we have some documentation on what should be exported?

**Comment by mpreiner** [comment]
> LGTM, just a few minor formatting issues.
> Do we have some documentation on what should be exported?

Ideally, we'd only need to export the public API and the parser. However, currently there is no clean separation between the public API and the main driver + parser. I initially removed all CVC4_PUBLIC and then added CVC4_EXPORT to classes/functions that are currently needed by the main driver + parsers. That means that the majority of CVC4_EXPORT in `src/` (excluding `src/api`) should be removed once we have a clean separation.

**Pull request merged**